FRA Parser

elipe17 opened the pull request:
## Summary of Changes
- Implemented FRA parser
- Remove some Elastic code to avoid creating an elastic doc for the model
- Created new Row classes
- Added unit tests
- Removed the use of, and storage of the `self.errors` dictionary. This is dictionary was keeping a reference to all parser errors created throughout the entirety of parsing in memory. Very , wasteful so I ran an analysis locally using the WA file with lots of errors to empirically validate the waste of memory. Attached below are two text files with the system stats docker reported for the `develop` branch and this branch while parsing the WA file. For this test case, the removal of the `errors` dictionary freed up ~256MB of memory. That is quite substantial when we are working with 2-4GB of memory.

[no_errors_dict.txt](https://github.com/user-attachments/files/19055273/no_errors_dict.txt)
[errors_dict.txt](https://github.com/user-attachments/files/19055274/errors_dict.txt)


Pull request closes #3474 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Run the tests or use `shell_plus` to duplicate the `parser_task` logic and query the DB for errors/models
2. Since the frontend is still being worked we can't submit a file that way

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] FRA Django model created
- [x] FRA parser implemented
- [x] Basic FRA validation handled in parser
- [x] New unit testing introduced for FRA files
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Create tanf exiter model and migration
- Update csv/xlsx decoders to handle empty/commented rows
- Create FRA row schema
- Update program manager to have refs to FRA schemas
- Update fra files to have more reasonable test cases
- Created new validators for FRA use cases

elipe17 committed:
- Remove elastic code
- Create fake class for FRA parser unitl elastic docs removed
- Create new Row type adn update row hashing and equal operators
- Force row data to be tuples
- Add parser to factory
- Update FRA header logic
- Add initial unit tests for FRA
- Add meaningful FRA test fieles

elipe17 committed:
- linting

elipe17 committed:
- removed all reference to the errors dictionary
- updated tests

elipe17 committed:
- linting

elipe17 committed:
- Update tests

elipe17 committed:
- Add more unit tests for FRA files
- Updated FRA parser to fail header validation if any value in the first row is not "truthy"

elipe17 committed:
- Rename row
- handle case in tuplerow to return range of values

elipe17 committed:
- alphabetize

elipe17 committed:
- Add tests for dataclasses
- Add ne method for dataclass

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 committed:
- testing base image

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 committed:
- Update migration
- Give model record type default
- Enable parsing from frontend

elipe17 committed:
- Add admin view of TE1 records

jtimpe commented on tdrs-backend/tdpservice/data_files/serializers.py:
```diff
@@ -124,6 +124,6 @@ def validate_section(self, section):
         if DataFile.Section.is_fra(section):
             user = self.context.get('user')
-            if not user.has_fra_access:
+            if not user.has_fra_access and not user.is_ofa_sys_admin:
```
```suggestion
            if not user.has_fra_access or not user.is_ofa_sys_admin:
```

ADPennington replied:
i get this error (on line 128) when submitting an FRA file as an admin and the first row is blank (and other rows not blank). just want to confirm that this the expected behavior?

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/fra/te1.py:
```diff
@@ -0,0 +1,55 @@
+        document=TANF_Exiter1Document,
+        fields=[
+            Field(
+                item="",
```
do we care about item numbers for these models?

jtimpe replied:
also the `record_type` doesn't get saved to the db
![image](https://github.com/user-attachments/assets/6d29ccb5-8bc9-40b7-ab71-cdbf68ec985c)

elipe17 replied:
Right, I need to actually remove the schema field because I have it set to a default value in the actual Django model. Nice catch!

elipe17 replied:
Sorry, im going crazy. If you pull the latest on this branch the record type field will be populated. I add real item numbers in the FRA error report generator PR :)

jtimpe reviewed (approved):
Looks really good! Everything is functional, just a couple nits about missing fields in the parsed object.

raftmsohani commented on tdrs-backend/tdpservice/parsers/decoders.py:
```diff
@@ -102,31 +102,38 @@ def _open_as_csv(self):
                 os.remove(self.local_file.name)
-                logger.info(f"Deleted tempory storage of csv file: {self.raw_file.name}")
+                assert os.path.exists(self.local_file.name) is False
         except Exception:
```
suggest adding either the error in the exception message, or add specific exceptions like FileNotFoundError, etc. I see you are checking the path exist, but I remember there are special cases where `os.path.exists` returns false positive

elipe17 replied:
I can't use the logger here unfortunately. It's stream is closed because the module goes out of scope by this point which then causes it to throw an exception. Thats why i added the assert instead.

raftmsohani commented on tdrs-backend/tdpservice/parsers/schema_defs/fra/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .te1 import te1
```
nit: use absolute path

raftmsohani reviewed (approved):
only two nits, otherwise looks good

elipe17 committed:
- Customize FRA field validators to generate cat1 errors
- Update tests based on new logic

elipe17 commented on tdrs-backend/tdpservice/parsers/row_schema.py:
```diff
@@ -265,10 +265,64 @@ def parse_and_validate(self, row: RawRow, generate_error):
+
+        NOTE: FRA (for the moment) needs all field based validators to produce category1 errors. This is the same exact
+        logic as RowSchema.run_field_validators, but we need to override it here to ensure that the error category is
+        correct.
```
This is the change I made to force all FRA errors to be cat1 while we implement cat4 logic. This was the easiest method so that I didn't have to write new validators or change the schema instances.

jtimpe replied:
haven't dug into this code too much, but would it be possible to instead override `generate_error` and pass it into the `RowSchema`? that way the logic doesn't have to be redefined, we just call a different version of the error generator

elipe17 committed:
- absolute paths

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

jtimpe commented on tdrs-backend/tdpservice/parsers/row_schema.py:
```diff
@@ -265,10 +265,64 @@ def parse_and_validate(self, row: RawRow, generate_error):
+                        errors.append(
+                            generate_error(
+                                schema=self,
+                                error_category=ParserErrorCategoryChoices.PRE_CHECK,
```
`ParserErrorCategoryChoices.PRE_CHECK` does mean the file status gets set to `REJECTED` - when we talked in office hours, i thought i remembered you saying that behavior was not desirable and instead we wanted to report the file status as `PARTIALLY_ACCEPTED` in the case of errors. has that changed?

elipe17 replied:
Correct, we do want it to be paritially accepted. However, this is better in the meantime as it is very strict and clear as opposed to accepted with errors. #3521 will address updating all of these validators to be cat4 and therefore force the status to partially accepted.

jtimpe reviewed (approved):
Changes are functional - approved though I did have one point of clarification and a small suggestion

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 committed:
Merge branch '3473-schema-field-update' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

elipe17 commented on tdrs-backend/tdpservice/search_indexes/models/fra.py:
```diff
@@ -0,0 +1,28 @@
+        on_delete=models.CASCADE,
+        related_name='tanf_exiter_1_parent'
+    )
+
```
Add STT object as first class citizen here

elipe17 replied:
ticket STT object on all datafile record models (maybe Region(s) too)

ticket let parser error know what value and fields generated it

raftmsohani commented on tdrs-backend/tdpservice/parsers/dataclasses.py:
```diff
@@ -87,6 +87,29 @@ def __hash__(self):
         """Return hash of data."""
         return hash(self.data)
 
+    def __eq__(self, value):
```
I wish python had a way to overload operators :)

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
- Added STT as a first class citizen on the TE1 model

elipe17 committed:
Revert "- Added STT as a first class citizen on the TE1 model"

This reverts commit 0e70ee93277459f00b0a4ebc306169a9f1895fa3.

ADPennington committed:
Merge branch 'develop' into 3474-fra-parser

ADPennington commented:
@elipe17 @lhuxraft @reitermb @victoriaatraft ive created some test files for this report. I'm dropping them below for our future reference. 

Since this PR introduces the first FRA error report, thought it would be helpful if we all had a baseline on how it looks and what its catching. 

[Test_CSV.csv](https://github.com/user-attachments/files/19216153/Test_CSV.csv)
[Test_XLSX.xlsx](https://github.com/user-attachments/files/19216154/Test_XLSX.xlsx)

edit: these are being tested for FY2025 Q3 😄

lhuxraft commented:
> @elipe17 @lhuxraft @reitermb @victoriaatraft ive created some test files for this report. I'm dropping them below for our future reference.
> 
> Since this PR introduces the first FRA error report, thought it would be helpful if we all had a baseline on how it looks and what its catching.
> 
> [Test_CSV.csv](https://github.com/user-attachments/files/19216153/Test_CSV.csv) [Test_XLSX.xlsx](https://github.com/user-attachments/files/19216154/Test_XLSX.xlsx)
> 
> edit: these are being tested for FY2025 Q2 😄

Thanks, Alex! Want to confirm if there's any testing you'd like UX to do on this ticket with these files?

ADPennington commented:
> > @elipe17 @lhuxraft @reitermb @victoriaatraft ive created some test files for this report. I'm dropping them below for our future reference.
> > Since this PR introduces the first FRA error report, thought it would be helpful if we all had a baseline on how it looks and what its catching.
> > [Test_CSV.csv](https://github.com/user-attachments/files/19216153/Test_CSV.csv) [Test_XLSX.xlsx](https://github.com/user-attachments/files/19216154/Test_XLSX.xlsx)
> > edit: these are being tested for FY2025 Q3 😄
> 
> Thanks, Alex! Want to confirm if there's any testing you'd like UX to do on this ticket with these files?

i recommend doing so as a pre-cursor to a larger discussion we'll want to have on the error report. also @elipe17 corrected my fiscal-cal math, so i've corrected the fiscal quarter on the comment above. 😄 cc: @klinkoberstar

elipe17 committed:
- Add none check in parse_value
- Move preparsing validators to base rowschema class
- dont allow default lists in class params
- Add preparsing validator for te1
- Add fixtures for ofa test cases
- Add test cases for ofa files
- Consolodate date validations into separate func

elipe17 committed:
- Update test given new val logic

elipe17 committed:
- linting

elipe17 committed:
- Update all SSNs to start with a 9 since that implies they are fake
- Corrected logic on field check

elipe17 committed:
- Correct checks on decoder tests

ADPennington commented on tdrs-backend/tdpservice/parsers/test/data/fra_empty_first_row.csv:
```diff
@@ -0,0 +1,14 @@
+202403,909175891
+
+# All data below has errors of some sort
+202301,9000865507
```
```suggestion
202301,0000865507
```
this one would capture the all zeros for area number, all zeros for the group number, and 10 digits 😄

ADPennington commented on tdrs-backend/tdpservice/parsers/fra_parser.py:
```diff
@@ -0,0 +1,97 @@
+        generate_error = make_generate_parser_error(self.datafile, 1)
+        self.num_errors += 1
+        error = generate_error(schema=None,
+                               error_category=ParserErrorCategoryChoices.PRE_CHECK,
```
ah this answers my earlier question 😄

ADPennington reviewed (approved):
LGTM @elipe17 🚀 notes below for reference @klinkoberstar @victoriaatraft @reitermb @lhuxraft 

- able to see the datafile object, file status, and parsing errors from DAC
- able to see the submission status and report in history 
- able to query te1 records in the db

Known issues as of this PR:
- error report not currently matching the version proposed in the KC draft [here](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/uAlDVLhlFIbJw6huUS2x/knowledge-center-drafts/fra-data-files-reporting-1). Please confirm which ticket is scoped for alignment with that design. 
- related to the above, need UX input on error messages. 
- duplicate records are currently being accepted into the db and are not raising an error (#3527)
- file status strictly accepted or rejected (#3527)
- total errors not populated yet. which ticket will implement this? also curious what level of aggregation we want to do here. Total errors or total errors per row? I'd be curious what end users prefer. 
- file renaming is not happening yet, and xlsx files are not download-able yet. whch ticket addresses this? 


other notes for future ticket considerations:
- with the single page submission and history view, it doesnt seem as valuable to have the most recent submission download button on the screen. 
- if the file has a blank first row, its rejected, but if the blank row is in the middle of the file, this doesnt raise an error, and the blank row isnt stored. Curious if we should re-consider if its okay for the file to be blank on the first row. its not really a problem unless there's nothing else to parse. This is something to monitor as part of early access testing before we make any decisions. 
- my test file, which was designed to catch a wide range of issues, has highlighted the possibility of getting multiple errors per row, which could be overwhelming (in a similar way to TANF/SSP error feedback). This is an important area for UX research to explore. We're already thinking about ways to consolidate the errors without losing all the detail (e.g. concatenating all possible issues/errors detected per row into an error code -- which would be accompanied by an explainer guide in the KC), but don't want to over-engineer at this early stage. 
- submission-related behavior:
    - sometimes the file status immediately changes, and other times it stays in pending. clicking search button yields the final file status. 
    - files with multiple sheets are rejected for unknown reason. if an issue, consider being explicit that this is why. 
    - If you upload a file with the wrong extension and then correct it, the new file will appear twice. It seems to upload fine but may be confusing


test file and error report included below:
[FY2025_Q3_FRA_WOTE_test.csv](https://github.com/user-attachments/files/19251456/FY2025_Q3_FRA_WOTE_test.csv)
[2025-Q2-Work Outcomes of TANF Exiters (6).xlsx](https://github.com/user-attachments/files/19251458/2025-Q2-Work.Outcomes.of.TANF.Exiters.6.xlsx)

lhuxraft commented:
> LGTM @elipe17 🚀 notes below for reference @klinkoberstar @victoriaatraft @reitermb @lhuxraft
> 
> * able to see the datafile object, file status, and parsing errors from DAC
> * able to see the submission status and report in history
> * able to query te1 records in the db
> 
> Known issues as of this PR:
> 
> * error report not currently matching the version proposed in the KC draft [here](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/uAlDVLhlFIbJw6huUS2x/knowledge-center-drafts/fra-data-files-reporting-1). Please confirm which ticket is scoped for alignment with that design.
> * related to the above, need UX input on error messages.
> * duplicate records are currently being accepted into the db and are not raising an error ([Enhance ParserError model to store field values that generated errors #3527](https://github.com/raft-tech/TANF-app/issues/3527))
> * file status strictly accepted or rejected ([Enhance ParserError model to store field values that generated errors #3527](https://github.com/raft-tech/TANF-app/issues/3527))
> * total errors not populated yet. which ticket will implement this? also curious what level of aggregation we want to do here. Total errors or total errors per row? I'd be curious what end users prefer.
> * file renaming is not happening yet, and xlsx files are not download-able yet. whch ticket addresses this?
> 
> other notes for future ticket considerations:
> 
> * with the single page submission and history view, it doesnt seem as valuable to have the most recent submission download button on the screen.
> * if the file has a blank first row, its rejected, but if the blank row is in the middle of the file, this doesnt raise an error, and the blank row isnt stored. Curious if we should re-consider if its okay for the file to be blank on the first row. its not really a problem unless there's nothing else to parse. This is something to monitor as part of early access testing before we make any decisions.
> * my test file, which was designed to catch a wide range of issues, has highlighted the possibility of getting multiple errors per row, which could be overwhelming (in a similar way to TANF/SSP error feedback). This is an important area for UX research to explore. We're already thinking about ways to consolidate the errors without losing all the detail (e.g. concatenating all possible issues/errors detected per row into an error code -- which would be accompanied by an explainer guide in the KC), but don't want to over-engineer at this early stage.
> * submission-related behavior:
>   
>   * sometimes the file status immediately changes, and other times it stays in pending. clicking search button yields the final file status.
>   * files with multiple sheets are rejected for unknown reason. if an issue, consider being explicit that this is why.
>   * If you upload a file with the wrong extension and then correct it, the new file will appear twice. It seems to upload fine but may be confusing
> 
> test file and error report included below: [FY2025_Q3_FRA_WOTE_test.csv](https://github.com/user-attachments/files/19251456/FY2025_Q3_FRA_WOTE_test.csv) [2025-Q2-Work Outcomes of TANF Exiters (6).xlsx](https://github.com/user-attachments/files/19251458/2025-Q2-Work.Outcomes.of.TANF.Exiters.6.xlsx)

@ADPennington Thanks so much for the detailed feedback! Looking at this ticket deployed in a11y, a couple questions:
- This ticket does not add error report functionality (that will be coming in #3419), but agree that error reports should not be generated at this stage. @elipe17 can you provide more detail?
- Total errors should be implemented as part of #3419 as well, but @elipe17 can you confirm this please? In original mockups, total errors was calculated by file but I can add this to the list of questions for early access testing.
- @reitermb I think we may have missed file renaming in the ACs. Per the mockups, files should be renamed to TANF_work_outcome.csv or TANF_work_outcome.xlsx after submission. @elipe17 Would you prefer I create a separate ticket or implement it as part of this one?
- I'll add removing the 'Download' button as another question to prompt for in early access testing and add it to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) for post-early access.
- Adding first row blanks to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) as well
- I'll add the multiple errors per row issue to [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) as well, but I'm curious to see how this manifest in the error report generation tickets (#3419) that follows this
- I'm also not able to reproduce changing file status in a11y, but curious if this will persist when error report generation is built out. I will also add these submisison behaviors to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking).

lhuxraft commented:
@elipe17 Doing some testing in a11y and noticing some issues:
- Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
- I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
- I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?

Example files I was using:
-[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx)
-[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)

cc @ADPennington

elipe17 commented:
> > LGTM @elipe17 🚀 notes below for reference @klinkoberstar @victoriaatraft @reitermb @lhuxraft
> > 
> > * able to see the datafile object, file status, and parsing errors from DAC
> > * able to see the submission status and report in history
> > * able to query te1 records in the db
> > 
> > Known issues as of this PR:
> > 
> > * error report not currently matching the version proposed in the KC draft [here](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/uAlDVLhlFIbJw6huUS2x/knowledge-center-drafts/fra-data-files-reporting-1). Please confirm which ticket is scoped for alignment with that design.
> > * related to the above, need UX input on error messages.
> > * duplicate records are currently being accepted into the db and are not raising an error ([Enhance ParserError model to store field values that generated errors #3527](https://github.com/raft-tech/TANF-app/issues/3527))
> > * file status strictly accepted or rejected ([Enhance ParserError model to store field values that generated errors #3527](https://github.com/raft-tech/TANF-app/issues/3527))
> > * total errors not populated yet. which ticket will implement this? also curious what level of aggregation we want to do here. Total errors or total errors per row? I'd be curious what end users prefer.
> > * file renaming is not happening yet, and xlsx files are not download-able yet. whch ticket addresses this?
> > 
> > other notes for future ticket considerations:
> > 
> > * with the single page submission and history view, it doesnt seem as valuable to have the most recent submission download button on the screen.
> > * if the file has a blank first row, its rejected, but if the blank row is in the middle of the file, this doesnt raise an error, and the blank row isnt stored. Curious if we should re-consider if its okay for the file to be blank on the first row. its not really a problem unless there's nothing else to parse. This is something to monitor as part of early access testing before we make any decisions.
> > * my test file, which was designed to catch a wide range of issues, has highlighted the possibility of getting multiple errors per row, which could be overwhelming (in a similar way to TANF/SSP error feedback). This is an important area for UX research to explore. We're already thinking about ways to consolidate the errors without losing all the detail (e.g. concatenating all possible issues/errors detected per row into an error code -- which would be accompanied by an explainer guide in the KC), but don't want to over-engineer at this early stage.
> > * submission-related behavior:
> >   
> >   * sometimes the file status immediately changes, and other times it stays in pending. clicking search button yields the final file status.
> >   * files with multiple sheets are rejected for unknown reason. if an issue, consider being explicit that this is why.
> >   * If you upload a file with the wrong extension and then correct it, the new file will appear twice. It seems to upload fine but may be confusing
> > 
> > test file and error report included below: [FY2025_Q3_FRA_WOTE_test.csv](https://github.com/user-attachments/files/19251456/FY2025_Q3_FRA_WOTE_test.csv) [2025-Q2-Work Outcomes of TANF Exiters (6).xlsx](https://github.com/user-attachments/files/19251458/2025-Q2-Work.Outcomes.of.TANF.Exiters.6.xlsx)
> 
> @ADPennington Thanks so much for the detailed feedback! Looking at this ticket deployed in a11y, a couple questions:
> 
> * This ticket does not add error report functionality (that will be coming in [Develop mechanism to generate FRA error reports #3419](https://github.com/raft-tech/TANF-app/issues/3419)), but agree that error reports should not be generated at this stage. @elipe17 can you provide more detail?
> * Total errors should be implemented as part of [Develop mechanism to generate FRA error reports #3419](https://github.com/raft-tech/TANF-app/issues/3419) as well, but @elipe17 can you confirm this please? In original mockups, total errors was calculated by file but I can add this to the list of questions for early access testing.
> * @reitermb I think we may have missed file renaming in the ACs. Per the mockups, files should be renamed to TANF_work_outcome.csv or TANF_work_outcome.xlsx after submission. @elipe17 Would you prefer I create a separate ticket or implement it as part of this one?
> * I'll add removing the 'Download' button as another question to prompt for in early access testing and add it to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) for post-early access.
> * Adding first row blanks to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) as well
> * I'll add the multiple errors per row issue to [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking) as well, but I'm curious to see how this manifest in the error report generation tickets ([Develop mechanism to generate FRA error reports #3419](https://github.com/raft-tech/TANF-app/issues/3419)) that follows this
> * I'm also not able to reproduce changing file status in a11y, but curious if this will persist when error report generation is built out. I will also add these submisison behaviors to our [change tracking](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/yfhDBBVu7eJEUBHaCqzz/fra-research/fra-mvp-early-access-user-test/post-early-access-change-tracking).

- Correct, when I initially published this PR the FRA page did not have submission history yet. The report exists because Jan wired it up to the existing backend endpoint which does not know how to delineate between TANF reports and FRA reports. That logic will happen in the ticket Lauren tagged.
- Total errors will also be calculated in the the ticket Lauren tagged for this bullet. It is "failing" for the same reason as above. It is just hooked to the original endpoint which doesn't have real knowledge of FRA yet.
- We definitely missed this AC. I suggest creating a new ticket.
- Yes, lets track this as it makes the page feel a little cluttered
- I chose to treat the first row being blank or with data other than the expected two column format as a "header" error. I was mostly worried about an stt having descriptions or something unexpected in the first row as a "header".
- Sounds good.
- This is a function of the asynchronous nature of the submission. Sometimes Celery queues the task and parses immediately and sometimes it sits on the queue for an extra second or two. This is something that will be handled by Jan's "spinner" PR since it will poll the submission history endpoint to avoid causing the user to "refresh/research" the page.

elipe17 commented:
> @elipe17 Doing some testing in a11y and noticing some issues:
> 
> * Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
> * I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
> * I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?
> 
> Example files I was using: -[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx) -[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)
> 
> cc @ADPennington

- I will test out large files.
- You see multiple file inputs on the FRA page?
- This is on the FRA page? The only way (that we know of) a file can get stuck in pending is if the Celery worker runs out of memory.

elipe17 commented:
> > @elipe17 Doing some testing in a11y and noticing some issues:
> > 
> > * Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
> > * I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
> > * I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?
> > 
> > Example files I was using: -[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx) -[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)
> > cc @ADPennington
> 
> * I will test out large files.
> * You see multiple file inputs on the FRA page?
> * This is on the FRA page? The only way (that we know of) a file can get stuck in pending is if the Celery worker runs out of memory.

- I tested both of the files. The issue with the .xlsx is the parser generates 900k errors for it because it thinks the data in the columns is mis-formatted. This could potentially be because the file used Excel functions to get their values. The .csv file fails because it is encoded as "UTF8-SIG" which is very similar to "UTF8-BOM".

elipe17 commented:
> > > @elipe17 Doing some testing in a11y and noticing some issues:
> > > 
> > > * Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
> > > * I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
> > > * I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?
> > > 
> > > Example files I was using: -[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx) -[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)
> > > cc @ADPennington
> > 
> > 
> > 
> > * I will test out large files.
> > * You see multiple file inputs on the FRA page?
> > * This is on the FRA page? The only way (that we know of) a file can get stuck in pending is if the Celery worker runs out of memory.
> 
> * I tested both of the files. The issue with the .xlsx is the parser generates 900k errors for it because it thinks the data in the columns is mis-formatted. This could potentially be because the file used Excel functions to get their values. The .csv file fails because it is encoded as "UTF8-SIG" which is very similar to "UTF8-BOM".

- Something very interesting is happening. When I submit the CSV file through the backend (i.e. programmatically) I see the utf8-sig issue. If I submit the same file through the frontend form fields it actually tells me the data is ASCII encoded but that the file only has 2 bytes of info. When I download what the frontend submitted I see the screenshot below. Im betting the frontend is converting the files to UTF8 for fra like it is for tanf which is not what we want.

![Screenshot 2025-03-15 at 8 23 48 AM](https://github.com/user-attachments/assets/1e3bf499-d397-4638-88dd-3344a11101a3)

elipe17 commented:
> > > > @elipe17 Doing some testing in a11y and noticing some issues:
> > > > 
> > > > * Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
> > > > * I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
> > > > * I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?
> > > > 
> > > > Example files I was using: -[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx) -[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)
> > > > cc @ADPennington
> > > 
> > > 
> > > 
> > > * I will test out large files.
> > > * You see multiple file inputs on the FRA page?
> > > * This is on the FRA page? The only way (that we know of) a file can get stuck in pending is if the Celery worker runs out of memory.
> > 
> > 
> > 
> > * I tested both of the files. The issue with the .xlsx is the parser generates 900k errors for it because it thinks the data in the columns is mis-formatted. This could potentially be because the file used Excel functions to get their values. The .csv file fails because it is encoded as "UTF8-SIG" which is very similar to "UTF8-BOM".
> 
> * Something very interesting is happening. When I submit the CSV file through the backend (i.e. programmatically) I see the utf8-sig issue. If I submit the same file through the frontend form fields it actually tells me the data is ASCII encoded but that the file only has 2 bytes of info. When I download what the frontend submitted I see the screenshot below. Im betting the frontend is converting the files to UTF8 for fra like it is for tanf which is not what we want.
> 
> ![Screenshot 2025-03-15 at 8 23 48 AM](https://private-user-images.githubusercontent.com/125676261/423095653-1e3bf499-d397-4638-88dd-3344a11101a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwNDE4NzgsIm5iZiI6MTc0MjA0MTU3OCwicGF0aCI6Ii8xMjU2NzYyNjEvNDIzMDk1NjUzLTFlM2JmNDk5LWQzOTctNDYzOC04OGRkLTMzNDRhMTExMDFhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNVQxMjI2MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjEyNTQ3NTg1M2Y2OTBjYWQyMDNkNmFhZWM4OWVjODFlMDNmNmRmMzhjMzRlNzI0ZDgwOWQyMmU5ZjhiN2E5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._ZayQJsz9R5pO2Xxz8Zpq5Y-IvBpYRFaDCgIUZechwA)

- For the xlsx file, I have confirmed that it is capturing the calcualted field as the "value". I logged what the parser thought the value was for some fields and saw this. You can see right where the value is an int and what it is and then changes to calculated field and integer.

![Screenshot 2025-03-15 at 8 45 03 AM](https://github.com/user-attachments/assets/44f2cea6-f630-4d32-a123-43116138209b)

elipe17 committed:
- a little extra handling for utf8 encoding types

lhuxraft commented:
> > > > > @elipe17 Doing some testing in a11y and noticing some issues:
> > > > > 
> > > > > * Wanted to test out how performance is doing with large files and I'm experiencing latency / performance issues when uploading large files (200K records). The 'Submit Report' functionality is fairly fast, but both downloading the error report and searching for that data while its presumably uploading after the fact takes a very long time with no indication to the user as to what is happening (~5-10 minutes). Note that CSVs seem to go a bit faster than XLSX but are still slow. Can we test these latency issues and/or create a future enhancement for adding a progress bar / spinner to these stages?
> > > > > * I'm allowed to select and upload multiple files at once for some reason - not sure if this is intended behavior?
> > > > > * I'm not yet able to test for a successful parse because the statuses just note pending. Is this intended behavior at this stage?
> > > > > 
> > > > > Example files I was using: -[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19253572/TEST_FRA_XLSX.xlsx) -[TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19253574/TEST_FRA_CSV.csv)
> > > > > cc @ADPennington
> > > > 
> > > > 
> > > > 
> > > > * I will test out large files.
> > > > * You see multiple file inputs on the FRA page?
> > > > * This is on the FRA page? The only way (that we know of) a file can get stuck in pending is if the Celery worker runs out of memory.
> > > 
> > > 
> > > 
> > > * I tested both of the files. The issue with the .xlsx is the parser generates 900k errors for it because it thinks the data in the columns is mis-formatted. This could potentially be because the file used Excel functions to get their values. The .csv file fails because it is encoded as "UTF8-SIG" which is very similar to "UTF8-BOM".
> > 
> > 
> > 
> > * Something very interesting is happening. When I submit the CSV file through the backend (i.e. programmatically) I see the utf8-sig issue. If I submit the same file through the frontend form fields it actually tells me the data is ASCII encoded but that the file only has 2 bytes of info. When I download what the frontend submitted I see the screenshot below. Im betting the frontend is converting the files to UTF8 for fra like it is for tanf which is not what we want.
> > 
> > ![Screenshot 2025-03-15 at 8 23 48 AM](https://private-user-images.githubusercontent.com/125676261/423095653-1e3bf499-d397-4638-88dd-3344a11101a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwNDE4NzgsIm5iZiI6MTc0MjA0MTU3OCwicGF0aCI6Ii8xMjU2NzYyNjEvNDIzMDk1NjUzLTFlM2JmNDk5LWQzOTctNDYzOC04OGRkLTMzNDRhMTExMDFhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNVQxMjI2MThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjEyNTQ3NTg1M2Y2OTBjYWQyMDNkNmFhZWM4OWVjODFlMDNmNmRmMzhjMzRlNzI0ZDgwOWQyMmU5ZjhiN2E5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._ZayQJsz9R5pO2Xxz8Zpq5Y-IvBpYRFaDCgIUZechwA)
> 
> * For the xlsx file, I have confirmed that it is capturing the calcualted field as the "value". I logged what the parser thought the value was for some fields and saw this. You can see right where the value is an int and what it is and then changes to calculated field and integer.
> 
> ![Screenshot 2025-03-15 at 8 45 03 AM](https://private-user-images.githubusercontent.com/125676261/423100334-44f2cea6-f630-4d32-a123-43116138209b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIxNzk2OTAsIm5iZiI6MTc0MjE3OTM5MCwicGF0aCI6Ii8xMjU2NzYyNjEvNDIzMTAwMzM0LTQ0ZjJjZWE2LWY2MzAtNGQzMi1hMTIzLTQzMTE2MTM4MjA5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxN1QwMjQzMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NTE4NzUyMDhkNjY3YjQzM2MxYzJhOTZiZjFiOTNkMWY5NmRmMjRkMWNhMDg1YzU4ZTJkOGNkMmQ0OTY4ZmQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.B9uw6IwrMSe_evLEF6IJDnfsWxCcmNbZENoqGey8_Ck)

@elipe17 Thanks so much for looking into this! Redeployed to a11y - on the errors I mentioned:
1. Being able to select and upload multiple files at once may have been a freak thing, because I can't reproduce this it on the latest deploy. Comfortable calling it resolved!
2. For the Pending issue, here's a screenshot of an example file that is still in Pending from Friday - the rest of the files I tested uploading were successfully rejected:
![image](https://github.com/user-attachments/assets/c0af2c79-3d9c-40c2-8428-06df46cc04c7)

Also, the encoding thing is really weird but thanks for fixing it! Took a moment to test the same Excel file if I replace the formula with purely the values and that seemed to fix the issue with the app getting stuck 🥳

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3474-fra-parser

lhuxraft reviewed (approved):
Should be all good now! Created #3556 as a follow-on for the additional AC we missed
